### DISCUSSION TITLE
Gas Estimation

### DISCUSSION THREAD

**Discussion by maxfundingrate** [zkSync-Community-Hub/zksync-developers#500] [General]
### Environment

Mainnet (and testnet)

### Provide a brief description of the functionality you're trying to implement.

Programmatic execution of transactions with proper gas estimations. 

### What is the specific issue or error you're encountering?

The contract function we're trying to call involves a loop over external contract function calls with try/catch blocks. The issue is that sometimes it seems like we run out of gas but there is no explicit out of gas error. `debug_traceTransaction` gives `execution reverted: 0x` an empty message where as `debug_traceCall` gives a proper output, suggesting out of gas issue.

### Can you share the error messages or logs you're receiving, if any?

See above.

### Have you made any recent changes to the contract before encountering this issue?

No.

### Are there any external libraries or contracts that your contract interacts with?

`web3py`

### Can you provide the relevant portions of your contract code where the issue is occurring?

Yes, I can add to private github repo.

### Have you tried to isolate the problem, and if so, what were the results?

Yes see above.

### What steps have you already taken to try to resolve the issue?

Increasing gas limits by 2x, 3x from the `zks_estimate_fee` method seems to resolve it temporarily. We also added a simple EWMA fit to historical gas estimations but we still get failed occasionally which impacts our business logic severely. Here is the function in python, we took elements from the `zksync2` repo as well:

```python
            account: LocalAccount = Account.from_key(self.private_key)
            nonce = self.zk_w3.zksync.get_transaction_count(
                account.address, ZkBlockParams.COMMITTED.value
            )
            gas_price = self.zk_w3.zksync.gas_price

            func_call = TxFunctionCall(
                chain_id=chain_id,
                nonce=nonce,
                from_=self.address,
                to=to,
                data=call_data,
                gas_limit=0,  # unknown at this state, will be replaced by estimate_gas
                gas_price=gas_price,
                max_priority_fee_per_gas=100_000_000,
            )
            estimate_gas = self.zk_w3.zksync.eth_estimate_gas(func_call.tx)
            
            chain_id = self.zk_w3.zksync.chain_id
            signer = PrivateKeyEthSigner(account, chain_id)

            func_call, estimate_gas = await self.zk_build_txn(
                account, chain_id, to, call_data
            )

            tx_712 = func_call.tx712(estimate_gas)
            signed_message = signer.sign_typed_data(tx_712.to_eip712_struct())
            msg = tx_712.encode(signed_message)
            tx_hash = self.zk_w3.zksync.send_raw_transaction(msg)
            tx_receipt = self.zk_w3.zksync.wait_for_transaction_receipt(
                tx_hash, timeout=240, poll_latency=0.5
            )
```

**Comment by bxpana** [comment]
Thanks for the post, we'll check this out and get back to you as soon as we can

**Reply by bxpana** [reply]
Just a heads up we are doing a deeper dive on things and hope to have an update for you soon. Appreciate the patience 😁

**Comment by petarTxFusion** [comment]
This code snippet looks good, could you share more information about contract and what are you sending in the calldata so we can take a better look what is happening in the background?

**Reply by maxfundingrate** [reply]
Hey @petarTxFusion yes happy to, I'll invite you to our repo. The particular function we're calling is `processQueuedRequests` with currently about 5-10 batched orders. You can see that the function does a lot of try/catch with external contract function calls, which might be a reason for the inaccurate gas estimates?

**Reply by bxpana** [reply]
@maxfundingrate thanks for sharing the repo. Just talked with @petarTxFusion and they are still looking into the issue. Hope to have an update for you soon.

**Reply by maxfundingrate** [reply]
Any updates? @bxpana @petarTxFusion

**Reply by bxpana** [reply]
Looks like it's something deeper than originally expected. Could you also invite me to the repo when you get a chance?

**Reply by maxfundingrate** [reply]
@bxpana yes just invited you

**Comment by mm-zk** [comment]
@maxfundingrate I don't have access to your contract, but here's a potential explanation, why things might be failing.

### How gas estimation works
We do a binary search, with different gas limits, to see whether the transaction would succeed or fail.
Unfortunately this might not work well for scenarios, where the execution flow drastically differs depending on amount of gas left.

Especially, as many libraries (and I think zkSync server too - I have to check) **increase** the amount of gas returned - which now might cause a different execution flow.

### What can go wrong if you're using try/catch.

`execution reverted: 0x` - suggests that you run out of gas in your 'main' contract (that was not part of try catch).


Imagine following code:
```
  try t.doSomething(gas: 10000) {
    try t.doSthElse(gas: 100000) {
    } catch {
     sstore(); // expensive op costs 8k.
    }
  } catch {
    // still successful.
    sstore(); // costs 8k;
    return;
  }
   
```
Now - it might happen, that you'll get the gas estimate of 9k here (where 'doSomething' fails quickly - as it sees it doesn't have enough gas - but you catch it in the catch and return ok). 
But suddenly, if you send 11k as gas limit, you'll trigger the different code flow, that would trigger 'doSthElse' - that will fail, but then the remaining gas might not be enough for sstore() operation - and you end up with 'execution reverted'.

**Reply by bxpana** [reply]
@maxfundingrate let us know if that ends up working for you or not.
